model/user/db.go

```diff
@@ -336,9 +341,55 @@ func GetOrCreateUser(userId, displayName, email, accessToken, refreshToken strin
 		return nil, errors.Wrapf(err, "decoding user '%s'", userId)
 
 	}
+
+	if err := setSlackInformation(env, u); err != nil {
+		return nil, errors.Wrapf(err, "setting slack information for user '%s'", userId)
+	}
+
 	return u, nil
 }
 
+func setSlackInformation(env evergreen.Environment, u *DBUser) error {
+	if u.Settings.SlackMemberId != "" {
+		// user already has a slack member id set
+		return nil
+	}
+	if u.EmailAddress == "" {
+		// we can't fetch the slack information without the user's email address
+		return errors.New("user has no email address")
+	}
+
+	// get the slack userID and memberId from the user's email
+	slackEnv := env.Settings().Slack
+	slackUser, err := slackEnv.Options.GetSlackUser(slackEnv.Token, u.EmailAddress)
+	if err != nil {
+		return errors.Wrapf(err, "getting Slack user with email address '%s'", u.EmailAddress)
+	}
+	if slackUser == nil {
+		grip.Error(message.Fields{
+			"message": fmt.Sprintf("Couldn't find slack user by email address", u.EmailAddress),
+			"userId":  u.Id,
+		})
+		return nil
+	}
+
+	slackFields := bson.M{
+		bsonutil.GetDottedKeyName(SettingsKey, userSettingsSlackMemberIdKey): slackUser.ID,
+	}
+	if slackUser.Name != "" {
+		slackFields[bsonutil.GetDottedKeyName(SettingsKey, userSettingsSlackUsernameKey)] = slackUser.Name
+	}
+
+	update := bson.M{"$set": slackFields}
+
+	if err := UpdateOne(bson.M{IdKey: u.Id}, update); err != nil {
```

Comment:
Can this update be combined with the one update above this? Seems like the Slack fields could just be set in L327 (contingent on `GetSlackUser` succeeding).

Reply:
done

Reply:
I moved it out because GetOrCreateUser doesn't do a find on the userID, so we can't know if the slackMemberId is set until it returns the user at the end.

Reply:
(so we would have to ping slack every time GetOrCreateUser is called, which I'm trying to avoid)